### ISSUE TITLE
Final Cut Pro FCPXML Support

### ISSUE THREAD

**Issue by orchetect** [orchetect/DAWFileKit#22]
## Parsing and Element Extraction

- [ ] `sync-clip`
  - currently their contents are not extracted with `extractElements()`
- [ ] `mc-clip`
  - currently their contents are not extracted with `extractElements()`
- [ ] Method to parse top-level story elements that are immediate children of `fcpxml`
- [ ] Detection for elements that are out of their containing clip's timecode bounds
  - write FCPXML utility method to detect if a clip or attribute (marker, caption, etc.) is out of bounds of the clip it's within
  - but also have a method that can recursively check up all the parent breadcrumbs to see if it's occluded from the main timeline
  - factor into `ExtractionSettings` as a Boolean property to filter out elements that are out-of-bounds and therefore not visible on the main timeline
- [ ] Change `absoluteStart` to `inTime`. Add `outTime` property to `context`, calculated from inTime + duration.

## Refactors

- [ ] Refactor FCPXML to remove `resources` parameter in inits and force them to parse exclusively from the XML, including parsing out resources when necessary

## Technical

- [ ] Add pretty `debugDescription` output for `Any*` enum cases, etc.
- [ ] Concurrency for FCPXML file parsing to increase performance
  - Each XML element's child notes could be parsed in parallel to speed up parsing. And that could happen with every element in the hierarchy.
- [ ] Make types `Sendable`

## To Investigate

- [ ] FCPXML `note` is a child element for clips, but an attribute for annotations

## Testing

- [ ] Unit test audio/video role attribute parsing & storage in structs

**Comment by orchetect** [comment]
A TON of work has been put into this thus far, but this remains very much a living and breathing in-progress effort. It's very close to reaching a point of satisfactory capability, stability, and accuracy. The goal now is to fine tune, work out some bugs, cover edge cases, and add unit testing to further improve its rigidity so it can be production-ready.

Currently, the parsing model is not complete (ie: does not implement 100% of the FCP XML 1.11 DTD). It may in future, but there is no compelling reason other than completeness. There exists a lot of ancillary elements possible in the DTD that are somewhat irrelevant to the main goal of DAWFileKit at present, which is markers extraction and interchange.

Also, the model is primarily designed to be read-only and is not yet capable of writing the model to disk as XML. That is a very possible addition to DAWFileKit in future, but will require a substantial refactor and is not a priority at this time.

DAWFileKit is currently capable of parsing and reasoning on:
- all resource types (asset, effect, format, locator, media, object-tracker, tracking-shape)
- all structure elements (library, event, project, sequence, spine)
- all story elements
  - all clip types (asset, audio, audition, clip, gap, mc-clip, ref-clip, sync-clip, title, video)
  - common annotations (markers, keywords, captions)

It's taken quite some time to start shaping the codebase and API into something more ergonomic and it's so much more effort than just writing an XML parser. FCPXML is a bit heady and takes time to get into to really understand and be able to reason on, and there's not a ton of great documentation out there.

**Comment by orchetect** [comment]
After a massive second-pass refactors, I'm getting close to stabilizing the API.

- The model takes the form of lightweight strongly-typed `XMLElement` wrappers.
- Careful protocolization and computed properties give read and write access to the XML without having to deal with the XML itself.
- No stored properties are utilized which allows for very performant XML traversal and data extraction.
- Advanced filtering methods and element extraction algorithms make it very easy to scrape data from the XML, such as extracting markers.
- Thorough support for calculating absolute time/timecode values for any element type from the standpoint of any parent container's timeline, including the main top-level timeline of course.
- Thorough support for extracting roles for elements including default roles and similar interpolations that Final Cut Pro itself makes but does not necessarily encode into the XML.
- Lots more improvements and features
- Improved scalability and maintainability

The goal is to have this pushed to a new release in the next week or so.

**Comment by orchetect** [comment]
The re-refactor is basically complete on `fcpxml-refactor` branch.

- Unit tests are passing
- There are **_substantial_** performance improvements for element extraction

There should be a forthcoming merge to main and release soon.

**Comment by orchetect** [comment]
Merged PR to main.
